danceradvent/public/articles/2015/7-medium-scale-part-5.pod

```diff
@@ -0,0 +1,208 @@
+interchangeable, you must not swap the C<PUT> and C<POST> verbs from the
+way I've defined them above. The HTTP spec says
+L<C<PUT> is idempotent|http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.1>, meaning
+that the browser is allowed to cache C<PUT> requests, suppressing
+subsequent identical calls if the request parameters do not change. In
+CRUD terms, this means that if you define "create" in terms of C<PUT>,
+the browser might not let you create two records with identical
+contents but different IDs, which means you risk losing data. The HTTP
+spec does not allow browsers to assume that identical C<POST> calls are
+idempotent, however, so we use that verb for CRUD's create operation,
+leaving C<PUT> = update.
+
+That's enough design philosophy; how does all that affect our Dancer
+code? We might decide that since we've almost got a REST API here that
+we might want to push it the rest of the way.
+
+Let's move the CRUD-like route handlers in C<lib/App/MajorFeature.pm> to
+C<lib/App/API/SubFeature.pm>, and also refine the programming interface
+a bit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_post;
+         get  '/'    => \&_get;
+         put  '/:id' => \&_put;
+         del  '/:id' => \&_del;
+     };
+ };
+
+My choice to use leading underscores in these module-internal route
+handler function names lets us parallel the Dancer DSL keyword naming
+scheme.
+
+You might instead prefer to make the HTTP-to-CRUD mapping more explicit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_create;
+         get  '/'    => \&_read;
+         put  '/:id' => \&_update;
+         del  '/:id' => \&_delete;
+     };
+ };
+
+As in the previous article in this series, the leading underscore
+convention saves us some aggravation here since C<delete> is a Perl
+keyword, and C<read> is a core Perl function.
+
+The only other thing to change here is that REST APIs normally return
+JSON or XML. I prefer JSON since that's directly usable by the
+JavaScript code that made the Ajax call to one of these APIs. One of the
+best features of Dancer is that automatic JSON encoding is built right
+in. We might define the C<_create()> function referenced above as:
+
+ sub _create {
+     my $params = request->params;
+     my $db     = get_db_conn;
+ 
+     if ($db->create_something($params->{arg1}, $params->{arg2})) {
+         return {
+             success => JSON::true,
+             id      => $db->last_insert_id(),
+         };
+     }
+     else {
+         return {
+             success => JSON::false,
+             error   => 'failed to create record: ' . $db->last_error(),
+         };
+     }
+ }
+
+In Dancer, a hash reference returned from a route handler is
+automatically JSON-encoded, and the HTTP response's content type is set
+to C<application/json>. Parsing and using that reply on the client side
+in JavaScript is therefore trivial.
```

Comment:
Only if the serializer is set. That also means you cannot return any HTML or it will break.

Additional comments:
1. You could use `send_as` to override the serializer for a single response.
2. Some structures cannot be represented as parameters (arrayref in JSON - because it has no keys) so you will need to decode that yourself using `request->data`. Some structures aren't represented well in parameters, such as a key that has an arrayref with a single item as its value. You will lose that subtly.